cloud/ec2.go

```diff
@@ -523,8 +523,86 @@ func (m *ec2Manager) getResources(ctx context.Context, h *host.Host) ([]string,
 	return resources, nil
 }
 
-// ModifyHost modifies a host according to the changes specified by a HostModifyOptions struct.
-func (m *ec2Manager) ModifyHost(ctx context.Context, h *host.Host, changes host.HostModifyOptions) error {
+// addTags adds or updates the specified tags in the client and db
+func (m *ec2Manager) addTags(ctx context.Context, h *host.Host, tags []host.Tag, resources []string) error {
+	createTagSlice := make([]*ec2.Tag, len(tags))
+	for i := range tags {
+		createTagSlice[i] = &ec2.Tag{Key: &tags[i].Key, Value: &tags[i].Value}
+	}
+	_, err := m.client.CreateTags(ctx, &ec2.CreateTagsInput{
+		Resources: aws.StringSlice(resources),
+		Tags:      createTagSlice,
+	})
+	if err != nil {
+		return errors.Wrapf(err, "error creating tags using client for '%s'", h.Id)
+	}
+	h.AddTags(tags)
+
+	return errors.Wrapf(h.SetTags(), "error creating tags in db for '%s'", h.Id)
+}
+
+// deleteTags removes the specified tags by their keys in the client and db
+func (m *ec2Manager) deleteTags(ctx context.Context, h *host.Host, keys, resources []string) error {
+	deleteTagSlice := make([]*ec2.Tag, len(keys))
+	for i := range keys {
+		deleteTagSlice[i] = &ec2.Tag{Key: &keys[i]}
+	}
+	_, err := m.client.DeleteTags(ctx, &ec2.DeleteTagsInput{
+		Resources: aws.StringSlice(resources),
+		Tags:      deleteTagSlice,
+	})
+	if err != nil {
+		return errors.Wrapf(err, "error deleting tags using client for '%s'", h.Id)
+	}
+	h.DeleteTags(keys)
+
+	return errors.Wrapf(h.SetTags(), "error deleting tags in db for '%s'", h.Id)
+}
+
+// setInstanceType changes the instance type in the client and db
+func (m *ec2Manager) setInstanceType(ctx context.Context, h *host.Host, instanceType string) error {
+	_, err := m.client.ModifyInstanceAttribute(ctx, &ec2.ModifyInstanceAttributeInput{
+		InstanceId: aws.String(h.Id),
+		InstanceType: &ec2.AttributeValue{
+			Value: aws.String(instanceType),
+		},
+	})
+	if err != nil {
+		return errors.Wrapf(err, "error changing instance type using client for '%s'", h.Id)
+	}
+
+	return errors.Wrapf(h.SetInstanceType(instanceType), "error changing instance type in db for '%s'", h.Id)
+}
+
+// setNoExpiration changes whether a host should expire
+func (m *ec2Manager) setNoExpiration(ctx context.Context, h *host.Host, noExpiration bool, resources []string) error {
```

Comment:
Can this go through createTags?

Reply:
It theoretically could, but createTags and the corresponding AddTags method on a host very explicitly does not overwrite any fields that are not marked as CanBeModified (and "expire-on" is one such system-set field that we don't want users to modify as a simple tag). Trying to keep user-created tags and system-created tags separate.